app/code/core/Mage/Usa/Model/Shipping/Carrier/UpsAuth.php

```diff
@@ -0,0 +1,109 @@
+<?php
+/**
+ * OpenMage
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available at https://opensource.org/license/osl-3-0-php
+ *
+ * @category   Mage
+ * @package    Mage_Usa
+ * @copyright  Copyright (c) 2006-2020 Magento, Inc. (https://www.magento.com)
+ * @copyright  Copyright (c) 2017-2024 The OpenMage Contributors (https://www.openmage.org)
+ * @license    https://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+
+/**
+ * UPS Authentication and Access Token handling
+ *
+ * @category   Mage
+ * @package    Mage_Usa
+ */
+class Mage_Usa_Model_Shipping_Carrier_UpsAuth extends Mage_Usa_Model_Shipping_Carrier_Abstract implements Mage_Shipping_Model_Carrier_Interface
+{
+    /**
+     * Cache key prefix for UPS API token
+     */
+    public const CACHE_KEY_PREFIX = 'ups_api_token_';
+
+    /**
+     * @return bool|string
+     */
+    public function getAccessToken(string $clientId, string $clientSecret, string $clientUrl)
+    {
+        $cacheKey = self::CACHE_KEY_PREFIX;
+        $cache = Mage::app()->getCache();
+        $result = $cache->load($cacheKey);
+        if ($result) {
+            return $result;
+        }
+
+        $headers = [
+            "Content-Type: application/x-www-form-urlencoded",
+            "x-merchant-id: $clientId",
+            "Authorization: Basic " . base64_encode("$clientId:$clientSecret"),
+        ];
+        $authPayload = http_build_query([
+            'grant_type' => 'client_credentials',
+        ]);
+        try {
+            $ch = curl_init();
```

Comment:
Don't we wanna use the Varien way?

Reply:
tried it and wasn't working for some reason, also in some of this APIs there was already curl calls so I went that route and worked correctly